batch2/batch/worker.py

```diff
@@ -243,114 +225,140 @@ def to_dict(self):
-    def path(self):
-        return self.name
+    def copy_command(src, dst):
+        if not dst.startswith('gs://'):
+            mkdirs = f'mkdir -p {shq(os.path.dirname(dst))};'
+        else:
+            mkdirs = ""
+        return f'{mkdirs} gsutil -m cp -R {shq(src)} {shq(dst)}'
 
-    async def delete(self):
-        await self.volume.delete()
+    copies = ' && '.join([copy_command(f['from'], f['to']) for f in files])
+    return f'{authenticate} && {copies}'
 
 
 class BatchPod:
-    async def _create_volumes(self):
-        log.info(f'creating volumes for pod {self.name}')
-        volumes = {}
-        for volume_spec in self.spec['spec']['volumes']:
-            name = volume_spec['name']
-            if volume_spec['empty_dir'] is not None:
-                volume = await EmptyDir.create(name)
-                volumes[name] = volume
-            elif volume_spec['secret'] is not None:
-                secret_name = volume_spec['secret']['secret_name']
-                path = f'{self.scratch}/secrets/{secret_name}'
-                secret = await Secret.create(name, path, self.secrets_data.get(secret_name))
-                volumes[name] = secret
-            else:
-                raise Exception(f'Unsupported volume type for {volume_spec}')
-        return volumes
-
     def __init__(self, worker, parameters, cpu_sem):
         self.worker = worker
-        self.spec = parameters['spec']
-        self.secrets_data = parameters['secrets']
-        self.output_directory = parameters['output_directory']
 
-        self.metadata = self.spec['metadata']
-        self.name = self.metadata['name']
+        job_spec = parameters['job_spec']
+
+        self.output_directory = parameters['output_directory']
+        self.name = parameters['name']
         self.token = uuid.uuid4().hex
         self.events = []
-        self.volumes = {}
+        self.scratch = f'/batch/pods/{self.name}/{self.token}'
+
+        input_files = job_spec.get('input_files')
+        output_files = job_spec.get('output_files')
+        pvc_size = job_spec.get('pvc_size')
+
+        # create volumes
+        self.volumes = []
+
+        copy_volume_mounts = []
+        main_volume_mounts = []
+
+        if job_spec.get('mount_docker_socket'):
+            main_volume_mounts.append('/var/run/docker.sock:/var/run/docker.sock')
+
+        if pvc_size or input_files or output_files:
+            v = Volume('io')
+            self.volumes.append(v)
+            volume_mount = 'io:/io'
+            main_volume_mounts.append(volume_mount)
+            copy_volume_mounts.append(volume_mount)
+
+        secrets = job_spec.get('secrets')
+        if secrets:
+            for secret in job_spec['secrets']:
+                host_path = f'{self.scratch}/{secret["name"]}'
+                populate_secret_host_path(host_path, secret['data'])
+                volume_mount = f'{host_path}:{secret["mount_path"]}'
+                main_volume_mounts.append(volume_mount)
```

Comment:
This will mount every secret in the main container. What happens if the secret should only be in the copy container?

Reply:
What secret only belongs in the copy container?  Right now it only gets the user's gsa-key (and the io volume).  The old code always mounted the gsa-key for the user (unless I'm misreading it).  Do you want to change that?